src/formatter.ts

```diff
@@ -92,34 +92,14 @@ export class JupyterlabNotebookCodeFormatter extends JupyterlabCodeFormatter {
     return codeCells;
   }
 
-  private getNotebookType() {
-    if (!this.notebookTracker.currentWidget) {
-      return null;
-    }
-
-    const metadata =
-      this.notebookTracker.currentWidget.content.model!.sharedModel.metadata;
-
-    if (!metadata) {
-      return null;
-    }
-
-    // prefer kernelspec language
-    if (
-      metadata.kernelspec &&
-      metadata.kernelspec.language &&
-      typeof metadata.kernelspec.language === 'string'
-    ) {
-      return metadata.kernelspec.language.toLowerCase();
-    }
-
-    // otherwise, check language info code mirror mode
-    if (metadata.language_info && metadata.language_info.codemirror_mode) {
-      const mode = metadata.language_info.codemirror_mode;
-      if (typeof mode === 'string') {
-        return mode.toLowerCase();
-      } else if (typeof mode.name === 'string') {
-        return mode.name.toLowerCase();
+  private getNotebookType(): string | null {
+    // Get the language from the current kernel spec
+    const sessionContext = this.notebookTracker.currentWidget?.sessionContext;
+    const kernelName = sessionContext?.session?.kernel?.name;
+    if (kernelName) {
+      const specs = sessionContext.specsManager.specs?.kernelspecs;
+      if (specs && kernelName in specs) {
+        return specs[kernelName]!.language;
```

Comment:
I think this is a fine approach to try, but I also think we should keep the old code as fallback. This is because there are kernels which will not have an associated session (these are kernels created outside of JupyterLab).

Reply:
Oh, that didn't even occur to me 😅 Will do.